### PR TITLE
[rcr] Reexport React.__COMPILER_RUNTIME.c or fallback to polyfill

### PR DESCRIPTION
**Pull request by poteto**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #31110
* #31055
* #31054
* #31053
* #31052
* __->__ #31051
* #31050

This PR updates the standalone `react-compiler-runtime` package to
either re-export `React.__COMPILER_RUNTIME.c` or to use a userspace
polyfill.

### CODE REVIEW

File: compiler/packages/react-compiler-runtime/src/index.ts [hunk 1]
```diff
@@ -19,22 +19,27 @@ const ReactSecretInternals =
...
+
+// Re-export React.c if present, otherwise fallback to the userspace polyfill for versions of React
+// < 19.
+export const c =
+  // @ts-expect-error
+  typeof React.c === 'function'
+    ? // @ts-expect-error
+      React.c
+    : function c(size: number) {
+        return React.useState<Array<unknown>>(() => {
```
Comment (josephsavona) [inline compiler/packages/react-compiler-runtime/src/index.ts:31]: aside: does `useState` reset when a fast refresh occurs that affects the given component? if not we could maybe use useMemo w no deps instead

File: compiler/packages/react-compiler-runtime/src/index.ts [hunk 2]
```diff
@@ -19,22 +19,27 @@ const ReactSecretInternals =
...
+          }
+          // This symbol is added to tell the react devtools that this array is from
+          // useMemoCache.
+          // @ts-ignore
+          $[$empty] = true;
+          return $;
+        })[0];
+      };
 
 export function $read(memoCache: MemoCache, index: number) {
```
Comment (josephsavona) [inline compiler/packages/react-compiler-runtime/src/index.ts:44]: nit: this is dead code, would be nice to clean it up later in the stack

File: compiler/packages/react-compiler-runtime/src/index.ts (hunk 1 above)
Comment (poteto) [reply compiler/packages/react-compiler-runtime/src/index.ts]: From what I can see it preserves state on FastRefresh, it only resets if the state variable got renamed

File: compiler/packages/react-compiler-runtime/src/index.ts (hunk 2 above)
Comment (poteto) [reply compiler/packages/react-compiler-runtime/src/index.ts:44]: will do!

Comment (poteto) [reply compiler/packages/react-compiler-runtime/src/index.ts:44]: https://github.com/facebook/react/pull/31110

File: compiler/packages/react-compiler-runtime/src/index.ts [hunk 3]
```diff
@@ -19,22 +19,27 @@ const ReactSecretInternals =
...
+    : function c(size: number) {
+        return React.useMemo<Array<unknown>>(() => {
+          const $ = new Array(size);
+          for (let ii = 0; ii < size; ii++) {
+            $[ii] = $empty;
+          }
+          // This symbol is added to tell the react devtools that this array is from
+          // useMemoCache.
+          // @ts-ignore
+          $[$empty] = true;
```
Comment (josephsavona) [inline compiler/packages/react-compiler-runtime/src/index.ts:39]: should we put this in `__DEV__`? not sure if this affects the JIT by changing the hidden class or not

**Review by josephsavona — APPROVED** [review]

File: compiler/packages/react-compiler-runtime/src/index.ts (hunk 3 above)
Comment (poteto) [reply compiler/packages/react-compiler-runtime/src/index.ts:39]: hmm probably wouldn't hurt but im not sure if this gets called enough for hidden classes to kick in?

**Pull request closed without merging**